### PR TITLE
Use docker initialization scripts for DB init

### PR DESCRIPTION
**Pull request by abitrolly**
The behavior is explained here https://hub.docker.com/_/postgres

>  After the entrypoint calls initdb to create the default postgres user and database, it will run any *.sql files, run any executable *.sh scripts, and source any non-executable *.sh scripts found in that directory to do further initialization before starting the service.

Also fixes #5980.

### CODE REVIEW

File: docker-compose.yml
Comment (ewjoachim) [inline docker-compose.yml:35]:
```suggestion
      - ./dev/example.sql.xz:/docker-entrypoint-initdb.d/example.sql.xz:ro
```
While we're at it's, it's free.

Comment (ewjoachim) [reply docker-compose.yml:35]: Do we want to add a named volume with the DB data ? This should help have the DB ready more often.

File: docker-compose.yml
Comment (ewjoachim) [inline docker-compose.yml:27]:
```suggestion
    image: postgres:10
```
While we're at it :)

File: Makefile
Comment (ewjoachim) [inline Makefile:145]: If we add a named volume, we'll need to delete it here.

File: docker-compose.yml
Comment (abitrolly) [reply docker-compose.yml]: > Do we also want to add a named volume with the DB data ? This should help have the DB ready more often.

I am not sure I understand why DB would be not ready in development process once initialized, and how named volume helps with that. For `make resetdb` it doesn't matter if it is a named volume or not.

Comment (abitrolly) [reply docker-compose.yml]: Oh, I see. The named volume that will persist already imported data.

Comment (ewjoachim) [reply docker-compose.yml]: yes

Comment (abitrolly) [reply docker-compose.yml]: I will address this in the thread below.

Comment (ewjoachim) [reply docker-compose.yml]: To elaborate: as of today, with an unnamed volume, if you delete (`docker-compose down`) and recreate your db container, the data is gone and you have to run `make initdb` (5 to 10 min) everytime. With a named volume, `docker-compose down` will leave the volume as-is, so the next time you run the stack, the DB is ready. If you need to reinit the db, you need to either delete the whole stack (`docker-compose down --volumes`) or delete the volume directly with `docker volume`.

File: Makefile
Comment (abitrolly) [reply Makefile:145]: So the named volume should hold the DB after it was imported and initialized. To me that extra redirection is not necessary, because `docker-compose down` doesn't kill the container contents. The only way DB content can be killed if somebody explicitly removes the containers explicitly with `rm` command. Why would we want to preserve the DB content in this case?

Comment (ewjoachim) [reply Makefile:145]: > The only way DB content can be killed if somebody explicitly removes the container with rm command

Or if the user uses `docker-compose down`.

As you want, really. I offered this as a suggestion. I know I do it in my projects usually and it works for me, but it doesn't have to be useful for everyone.

Comment (ewjoachim) [reply Makefile:145]: To me `docker compose down` means 'free all useless disk space taken by containers of this project' (and `docker compose down -v` means free all disk space taken by containers of this project'). I like being able to remove quickly everything except the thing that would be long to recreate.

Comment (abitrolly) [reply Makefile:145]: Can you confirm that `docker-compose down` really removes the containers? Because in my setup it doesn't happen. I will check out once more right now, but I am pretty sure the behavior is that compose just issues `stop` and not `rm`.

Or do you mean that `docker-compose down` cleans up volumes, and this is what you want? But then you will need to reset the DB and run migrations anyway.

Comment (abitrolly) [reply Makefile:145]: I was wrong. The compose setup survives reboot, but all containers are removed when `down` is invoked. So we need a named volume.

Comment (ewjoachim) [reply Makefile:159]: yep. `docker-compose down = docker-compose stop + docker-compose rm`, roughly

File: docker-compose.yml
Comment (abitrolly) [reply docker-compose.yml]: Added the named volume.

File: Makefile
Comment (abitrolly) [reply Makefile:159]: Added named volume removal with explicit `docker` command, because `docker-compose` doesn't support this https://stackoverflow.com/questions/45511956/remove-a-named-volume-with-docker-compose

**Comment by abitrolly** [comment]
Added the named volume and rebased. PTAL.

File: Makefile
Comment (ewjoachim) [inline Makefile:158]: Thinking about it, I realize `warehouse_pgdata` is actually `${COMPOSE_PROJECT_NAME:-$(basename $(pwd))}_pgdata`. Chance are people didn't change COMPOSE_PROJECT_NAME and their folder is named warehouse but this is just an assumption.

(note that `${COMPOSE_PROJECT_NAME:-$(basename $(pwd))}_pgdata` is the bash syntax, I don't know enough Makefile syntaxt to say what the equivalent would be)

Alternatives include specifying a specific `name` in the volume definition to avoid the compose project name being appended.

I realize this sucks a little bit :( I hadn't foreseen that deriving the volume name would be non-trivial until I saw this diff.

Comment (abitrolly) [reply Makefile:158]: Yea, well, need to test this with the `make` first.

Comment (ewjoachim) [reply Makefile:158]: Here's the make syntax to do it (ok I cheated)
```
$(shell python -c 'import os;print(os.getenv("COMPOSE_PROJECT_NAME", os.getcwd().rsplit("/", 1)[-1]) + "_pgdata")')
```

Comment (abitrolly) [reply Makefile]: I used standard `make` functions from http://www.gnu.org/software/make/manual/html_node/File-Name-Functions.html

Comment (abitrolly) [reply Makefile]: See the last commit.

**Review by ewjoachim — APPROVED** [review]

**Comment by abitrolly** [comment]
Interesting. The changes are approved, but workflow still doesn't run.

**Comment by ewjoachim** [comment]
Yes, I'm a pypi moderator, but I'm not an admin of the repo, I don't have the right to unblock the workflow.

**Comment by FFY00** [comment]
You need commit access.

**Comment by ewjoachim** [comment]
I know :) But that's not something I have as of today :)

**Review by ewdurbin — COMMENTED** [review]
Two questions:

1) does this handle decompressing the example.sql.xz file?
2) is there a mechanism to re-intialize, or does the Postgres container require being destroyed/recreated?

**Comment by abitrolly** [comment]
> does this handle decompressing the example.sql.xz file?

Yes, here.

https://github.com/docker-library/postgres/blob/740bb6d8d44a955df42d88fd5660c7043340db90/10/stretch/docker-entrypoint.sh#L168

> is there a mechanism to re-intialize, or does the Postgres container require being destroyed/recreated?

`DROP database` then load the dump with .xz as before, but that's not how containers are supposed to work.

**Comment by abitrolly** [comment]
I probably need to clarify the effects of this PR.

1. The DB initialization with creating `warehouse` DB and uploading initial data dump happens automattically when the `db` container starts for the first time, not waiting for the user to issue `initdb`. So when `db` becomes online, the only thing that needs to be run are migrations.
2. Because initial dump is already imported, `initdb` now doesn't kill the database, allowing to run migrations multiple times safely. Actually, it can now be renamed to `updatedb`.
3. `resetdb` command is added to do explicit resets to the state before migrations. This allows to better control for debugging migrations. The `resetdb` reuses the DB initialization mechanism from 1. to avoid maintenance of parallel SQL code that does the same thing. The DB initialization mechanism is triggered when the container starts for the first time and there is no existing database on disk. We need to stop or kill the container to clear the volume without unexpected errors. It is possible to just stop the `db` container, but that signals that `db` container contents outside volumes is actively modified, which is not the case here.

@ewdurbin does that makes things look better? I can insert the SQL code back to `resetdb` if you need it, and then the separate PR will be needed to address https://github.com/pypa/warehouse/pull/5980

**Comment by abitrolly** [comment]
So where we go from here?

**Comment by di** [comment]
While this seems like a nice feature of the postgres image, I'm not sure I like that database initialization is now split.

Ideally _all_ database initialization would be done automatically here, and we wouldn't need `initdb` at all, which means that we keep the dev database up-to-date with every new migration. This would require something like https://github.com/pypa/warehouse/pull/7496 but also we'd need to make the dump much smaller, so as to not introduce a large file into the git history every time there is a migration.

**Comment by abitrolly** [comment]
Database initialization is done automatically here. I could rename that to `migratedb` target as it is actually what the `initdb` target does at the moment. The only blocker is the line `docker-compose run web python -m warehouse sponsors populate-db`. I am not sure yet if it can be replaced with static import/export.

**Comment by abitrolly** [comment]
Okay. Sponsors line is a technical debt from https://github.com/pypa/warehouse/pull/9512/files#diff-b912d97d6c38546cdc81c0c2cc35833412c3b8ccd46a62ac5767f030a35bef73 Because there are not pathways to Alembic migrations for Django users.

**Comment by abitrolly** [comment]
Finally. 3 years later, but it is done. I would consider learning from Godot best practices for supporting the community that sends patches.

**Pull request closed without merging**
